File: src/main/java/com/vc/medium/L323.scala
package com.vc.medium

object L323 {
  def countComponents(n: Int, edges: Array[Array[Int]]): Int = {
    val pArr = new Array[Int](n)
    pArr.indices.foreach(i => pArr(i) = i)
    def find(x: Int): Int = {
      if(pArr(x) == x) return x
      find(pArr(x))
    }
    edges.foreach(edge => {
      val x = find(edge(0))
      val y = find(edge(1))
      pArr(x) = y
    })
    pArr.indices.foreach(i => if(pArr(i) != i) pArr(i) = find(pArr(i)))
    pArr.distinct.length
  }
}

File: src/main/java/com/vc/medium/L323.java
package com.vc.medium;

class L323 {
    public int countComponents(int n, int[][] edges) {
        int[] parent = new int[n];

        for(int i = 0; i < parent.length; i++) parent[i] = i;

        int components = n;
        for(int[] edge: edges) {
            int fromParent = find(parent, edge[0]);
            int toParent = find(parent, edge[1]);
            if(fromParent == toParent) continue;
            parent[fromParent] = toParent;
            components--;
        }

        return components;
    }

    private int find(int[] parent, int x) {
        if(parent[x] == x) return x;
        else return parent[x] = find(parent, parent[x]);
    }
}